src/v/archival/ntp_archiver_service.cc

```diff
@@ -851,27 +870,108 @@ ss::future<cloud_storage::upload_result> ntp_archiver::upload_segment(
       [this]() { return upload_should_abort(); },
     };
 
-    auto reset_func =
-      [this,
-       candidate]() -> ss::future<std::unique_ptr<storage::stream_provider>> {
-        return ss::make_ready_future<std::unique_ptr<storage::stream_provider>>(
-          std::make_unique<storage::concat_segment_reader_view>(
-            candidate.sources,
-            candidate.file_offset,
-            candidate.final_file_offset,
-            _conf->upload_io_priority));
+    // This struct wraps a stream object and exposes the stream_provider
+    // interface for compatibility with the remote object API.
+    struct stream_wrapper final : public storage::stream_provider {
+        stream_wrapper(const stream_wrapper&) = delete;
+        stream_wrapper(stream_wrapper&&) = default;
+        stream_wrapper& operator=(const stream_wrapper&) = delete;
+        stream_wrapper& operator=(stream_wrapper&&) = default;
+
+        ~stream_wrapper() override = default;
+
+        explicit stream_wrapper(ss::input_stream<char> s)
+          : stream(std::move(s)) {}
+
+        std::optional<ss::input_stream<char>> stream;
+        ss::input_stream<char> take_stream() override {
+            vassert(stream.has_value(), "no stream to take");
+            auto s = std::move(stream.value());
+            stream = std::nullopt;
+            return s;
+        }
+        ss::future<> close() override {
+            if (stream.has_value()) {
+                co_await stream->close();
+            }
+            co_return;
+        }
+    };
+
+    auto wrapper = std::make_unique<stream_wrapper>(
```

Comment:
We can only do this for the first call to `reset_func` when we have a stream as input from the fanout when we first opened the file, one stream is used for this upload and the other for the index creation. The second time onwards we need to read the file again.

Reply:
The problem with the fanout stream is that it's supposed to be used once. You can't really reuse it. But you can detach individual out streams from it. So, I guess the following approach could be taken:

- on first attempt the fanout can be used to scan the file. one stream is for segment upload, another to scan segment to generate the index in memory;
- if segment can't be uploaded and retry is required the stream has to be closed, in this case another stream which is used to generate segment will be working just fine. if index generation failed the whole operation has to be canceled and retried as a whole (some exception that can't be retried has to be thrown, I guess);
- upon retry the stream has to be opened normally (without fanout), just read it from the beginning;

Reply:
Maybe it makes sense to just upgrade redpanda indexes to be the same as cloud storage indexes.

Reply:
Another approach would be to use fanout stream to generate index without sending it until the segment is successfully uploaded. Retry should just recreate index and fanout streams. You can't upload index until the segment is fully scanned anyway. I think this is how it works in the read path.

Reply:
> on first attempt the fanout can be used to scan the file. one stream is for segment upload, another to scan segment to generate the index in memory;
if segment can't be uploaded and retry is required the stream has to be closed, in this case another stream which is used to generate segment will be working just fine. if index generation failed the whole operation has to be canceled and retried as a whole (some exception that can't be retried has to be thrown, I guess);
upon retry the stream has to be opened normally (without fanout), just read it from the beginning;

This is roughly what we do right now, two streams are opened, one is used for index generation and the other for segment upload. Both streams are kept in the same lifetime as the upload futures.

If the segment upload fails we close the fanout stream and retry with a new (non-fanout) normal file stream. 

If the index generation fails that is non recoverable. If the index upload fails, we are uploading the iobuf so the stream is not needed again, we just reupload the buffer.

However it often seems to run into problems when the system is shutting down. I assumed it would be an easy to track issue but it seems it is hard to figure out where the invariants break down (most probably a lifetime issue).

I can try the second approach, but I think it might run into similar issues as the current implementation faces. 

I do think I overlooked one thing from the second approach in the current implementation, that sometimes we may end up uploading the index but not the segment, which will later not be correctly cleaned up in case of topic deletion etc.

Reply:
I see a couple of issues here, first is that the method I used for closing the stream to upload segment sometimes missed closing it, causing the `parent must be closed` assertion triggered in segment reader handle.

The other is something I think might be a corner case in the input stream fanout https://github.com/redpanda-data/redpanda/issues/9779 where one stream closes early and the other has read some data at that point. This often happens in ducktape in my testing when redpanda is shutting down. The segment upload stream shuts down early, as `remote` does not upload if its gate is closed, and at the same time index creation stream has progressed a bit, thus triggering the condition for assert.

I will try to create PR for the workaround mentioned in https://github.com/redpanda-data/redpanda/issues/9779